lib/model/query/datasets.js

```diff
@@ -0,0 +1,236 @@
+    ON CONFLICT ON CONSTRAINT datasets_name_projectid_unique 
+    DO UPDATE SET "revisionNumber" = datasets."revisionNumber" + 1
+    RETURNING *
+  )
+  ${raw(withDsDefsCTE ? ', ds_defs AS (' : '')}
+    INSERT INTO dataset_form_defs
+    SELECT ds.id, ${dataset.aux.formDefId} FROM ds
+    ON CONFLICT ON CONSTRAINT dataset_form_defs_datasetid_formdefid_unique 
+    DO NOTHING
+  ${raw(withDsDefsCTE ? '),' : '')}
+`;
+
+const _createOrMerge = (dataset, fields, acteeId) => (isNilOrEmpty(fields) ? _insertDatasetDef(dataset, acteeId, false) : sql`
+${_insertDatasetDef(dataset, acteeId, true)}
+fields("propertyName", "formDefId", path) AS (VALUES      
+  ${sql.join(fields.map(p => sql`( ${sql.join([p.aux.propertyName, p.formDefId, p.path], sql`,`)} )`), sql`,`)}
+),
+dsProperties AS (
+    INSERT INTO ds_properties 
+    SELECT nextval(pg_get_serial_sequence('ds_properties', 'id')), fields."propertyName", ds.id FROM fields, ds
+    ON CONFLICT  ON CONSTRAINT ds_properties_name_datasetid_unique
+    DO NOTHING 
+    RETURNING ds_properties.id, ds_properties.name, ds_properties."datasetId"
+),
+dsPropertiesAll AS (
+    (SELECT dsProperties.id "dsPropertyId", dsProperties.name "propertyName", fields."formDefId", fields.path FROM fields
+    JOIN dsProperties ON fields."propertyName" = dsProperties.name)
+    UNION 
+    (SELECT ds_properties.id "dsPropertyId", ds_properties.name "propertyName", fields."formDefId", fields.path FROM fields
+    JOIN ds_properties ON fields."propertyName" = ds_properties.name)
+)
+INSERT INTO ds_property_fields
+SELECT "dsPropertyId", "formDefId"::integer, path FROM dsPropertiesAll
+ON CONFLICT ON CONSTRAINT ds_property_fields_dspropertyid_formdefid_path_unique 
+DO NOTHING
+`);
+
+const _getByIdSql = ((fields, datasetId) => sql`
+       SELECT
+           ${fields}
+       FROM
+           datasets
+       LEFT OUTER JOIN ds_properties ON
+           datasets.id = ds_properties."datasetId"
+       LEFT OUTER JOIN ds_property_fields ON
+           ds_properties.id = ds_property_fields."dsPropertyId"
+       WHERE datasets.id = ${datasetId}
+    `);
+
+const _getByNameSql = ((fields, datasetName, projectId) => sql`
+    SELECT
+        ${fields}
+    FROM
+        datasets
+    LEFT OUTER JOIN ds_properties ON
+        datasets.id = ds_properties."datasetId"
+    LEFT OUTER JOIN ds_property_fields ON
+        ds_properties.id = ds_property_fields."dsPropertyId"
+    WHERE datasets.name = ${datasetName}
+      AND datasets."projectId" = ${projectId}
+ `);
+
+
+// Creates or merges dataset, properties and field mapping.
+// Expects dataset:Frame auxed with `formDefId` and array of field:Frame auxed with `propertyName`
+// Note from Kathleen: I apologize at how hacky it is to get the provisioned actee ID into this
+// and get the created dataset out again...
+const createOrMerge = (dataset, fields) => ({ all, Actees, Datasets, Projects }) =>
```

Comment:
Revisit after merge: provision actee within SQL so this code is less complicated and faster.